SingularityService/src/main/java/com/hubspot/singularity/mesos/SingularityMesosSchedulerClient.java

```diff
@@ -0,0 +1,433 @@
+import org.apache.mesos.v1.Protos.ExecutorID;
+import org.apache.mesos.v1.Protos.Filters;
+import org.apache.mesos.v1.Protos.FrameworkID;
+import org.apache.mesos.v1.Protos.FrameworkInfo;
+import org.apache.mesos.v1.Protos.KillPolicy;
+import org.apache.mesos.v1.Protos.Offer;
+import org.apache.mesos.v1.Protos.OfferID;
+import org.apache.mesos.v1.Protos.TaskID;
+import org.apache.mesos.v1.scheduler.Protos.Call;
+import org.apache.mesos.v1.scheduler.Protos.Call.Accept;
+import org.apache.mesos.v1.scheduler.Protos.Call.Acknowledge;
+import org.apache.mesos.v1.scheduler.Protos.Call.Builder;
+import org.apache.mesos.v1.scheduler.Protos.Call.Decline;
+import org.apache.mesos.v1.scheduler.Protos.Call.Kill;
+import org.apache.mesos.v1.scheduler.Protos.Call.Message;
+import org.apache.mesos.v1.scheduler.Protos.Call.Reconcile;
+import org.apache.mesos.v1.scheduler.Protos.Call.Request;
+import org.apache.mesos.v1.scheduler.Protos.Call.Shutdown;
+import org.apache.mesos.v1.scheduler.Protos.Call.Type;
+import org.apache.mesos.v1.scheduler.Protos.Event;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.inject.Inject;
+import com.google.inject.name.Named;
+import com.google.protobuf.ByteString;
+import com.hubspot.singularity.SingularityMainModule;
+import com.hubspot.singularity.config.MesosConfiguration;
+import com.hubspot.singularity.config.SingularityConfiguration;
+import com.hubspot.singularity.config.UIConfiguration;
+import com.hubspot.singularity.resources.UiResource;
+import com.mesosphere.mesos.rx.java.AwaitableSubscription;
+import com.mesosphere.mesos.rx.java.MesosClient;
+import com.mesosphere.mesos.rx.java.MesosClientBuilder;
+import com.mesosphere.mesos.rx.java.SinkOperation;
+import com.mesosphere.mesos.rx.java.SinkOperations;
+import com.mesosphere.mesos.rx.java.protobuf.ProtobufMesosClientBuilder;
+import com.mesosphere.mesos.rx.java.util.UserAgentEntries;
+
+import rx.Observable;
+import rx.subjects.PublishSubject;
+import rx.subjects.SerializedSubject;
+
+/**
+ * Implementation of communication with the Mesos Master. To correctly close the
+ * connection when done you need to call {@link #close()}.
+ * <p>
+ * http://mesos.apache.org/documentation/latest/scheduler-http-api/
+ */
+public class SingularityMesosSchedulerClient {
+  private static final Logger LOG = LoggerFactory.getLogger(SingularityMesosSchedulerClient.class);
+
+  private final SingularityConfiguration configuration;
+  private final MesosConfiguration mesosConfiguration;
+  private final String singularityUriBase;
+
+  private SerializedSubject<Optional<SinkOperation<Call>>, Optional<SinkOperation<Call>>> publisher;
+  private FrameworkID frameworkId;
+  private AwaitableSubscription openStream;
+  private Thread subscriberThread;
+
+  @Inject
+  public SingularityMesosSchedulerClient(SingularityConfiguration configuration, @Named(SingularityMainModule.SINGULARITY_URI_BASE) final String singularityUriBase) {
```

Comment:
Would it make sense to have this implement AutoCloseable?

Reply:
My eventual thought for this is to have it be able to fix #273 , meaning the same singleton scheduler client could subscribe again or renegotiate its connection. The client isn't ever used in a single try-with-resource type of scope so the only closing that would have to be done is on shutdown. Also, the dropwizard guicier we use here doesn't have the auto-close-singleton-closeables bits that the other version does